### PR TITLE
[Tests] Functional Tests Suite Overhaul

### PR DESCRIPTION
**Pull request by random-zebra**
This is a collection of updates and improvements over the RegTest functional test suite.
Main changes:

### PoS blockchain cache
A new cache is created. It contains a 330-blocks chain (re using the regular "pow" chain for the first 200 blocks) with 8 nodes and some preminted zerocoin to speed up the tests. It is used in all pos/zerocoin tests.

### RPC fixes - additions
Additional fields in `spendzerocoin`, `spendrawzerocoin`, `spendzerocoinmints`, and `createrawzerocoinstake` specific for regtest, in order to make the tests more complete.
Lower required zerocoin confirmations to 10 on regtest.

### Refactor PIVX specific functions in test_framework / blocktools / utils
Unify code used to manually assemble PoS blocks (`get_prevouts`, `stake_block`). Add utility functions to create pos coinbases, spend utxos, generate pos blocks, manage sporks.

### Unify and complete broken tests
Group zerocoin spend tests in `zerocoin_spends`. Group Fake-Stake tests in `mining_pos_fakestake`. Fix `wallet_reorg_stake` and `zerocoin_wrapped_serials`.

Also updates (and re-sorts) the test runner list with the average time that each test requires.

### CODE REVIEW

**Comment by furszy** [comment]
Cool PR, this looks pretty nice. Finally decreasing the tests time, travis will love you.

Will start reviewing it in the upcoming days 👌

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3484,37 +3490,38 @@ UniValue spendzerocoin(const UniValue& params, bool fHelp)
     if(sporkManager.IsSporkActive(SPORK_16_ZEROCOIN_MAINTENANCE_MODE))
         throw JSONRPCError(RPC_WALLET_ERROR, "zPIV is currently disabled due to maintenance.");
 
-    EnsureWalletIsUnlocked();
```
Comment (Fuzzbawls) [inline src/wallet/rpcwallet.cpp]: should probably keep this check, as a locked wallet will ultimately fail `CWallet::SpendZerocoin()` anyways.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3548,61 +3555,53 @@ UniValue spendzerocoinmints(const UniValue& params, bool fHelp)
...
         throw JSONRPCError(RPC_WALLET_ERROR, "zPIV is currently disabled due to maintenance.");
 
-    std::string address_str = "";
-    if (params.size() > 1) {
-        RPCTypeCheck(params, boost::assign::list_of(UniValue::VARR)(UniValue::VSTR));
-        address_str = params[1].get_str();
-    } else
-        RPCTypeCheck(params, boost::assign::list_of(UniValue::VARR));
-
-    EnsureWalletIsUnlocked();
```
Comment (Fuzzbawls) [inline src/wallet/rpcwallet.cpp]: should probably keep this check, as a locked wallet will ultimately fail `CWallet::SpendZerocoin()` anyways.

**Comment by random-zebra** [comment]
@Fuzzbawls That check has not been deleted deliberately.
It was just moved inside `DoZpivSpend` which is called by
```
- spendzerocoin
- spendrawzerocoin
- spendzerocoinmints
```
and does the actual interaction with the wallet.

What those RPC functions do, before calling `DoZpivSpend`, is just parameter checks and/or data-structures set up.
Those can be done with a locked wallet and, imo, is even better to do those before checking (e.g. no reason to ask the user to unlock his wallet if, say, the provided destination address is invalid).

**Comment by Fuzzbawls** [comment]
@random-zebra right, i see it now. missed it's inclusion in `DoZpivSpend` initially.

**Pull request merged**